[CI/Build]: Update docker specification to be compatible with latest vLLM OpenAI server

hickeyma opened the pull request:
Updated the docker specification as follows:

- Removed redundant commands
- Updated requirements to latest packages
- Add build args to make build configurable for different vLLM images
- Added example build script

Partial #594

Tested by:
- Building image using latest vLLM image (`b8ed18f4b73f73f68fb238533a2ac1d61e471f50`) in AWS ECR registry: `public.ecr.aws/q9t5s3a7/vllm-ci-test-repo`
- Run the container which exposese the vLLM OpenAI server
- Test with `curl` command as per https://docs.lmcache.ai/getting_started/quickstart/offload_kv_cache.html

**Note to reviewers:** This is 1st step in providing pipeline for nightly build images as part of #594. There will be follow on PRs for centralizing requirement files and then pipeline file(s).

hickeyma committed:
Update docker specification to be compatible with latest vLLM OpenAI server

Updated the docker specification as follows:

- Removed redundant commands
- Updated requirements to latest packages
- Add build args to make build configurable for different vLLM images
- Added example build script

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

sammshen commented on docker/example_run.sh:
```diff
@@ -9,6 +19,5 @@ docker run --runtime nvidia --gpus all \
-    serve mistralai/Mistral-7B-Instruct-v0.2 --kv-transfer-config \
-    '{"kv_connector":"LMCacheConnector","kv_role":"kv_both"}' \
-    --enable-chunked-prefill false
+    serve $HF_MODEL_NAME --kv-transfer-config \
```
Great catch on the deprecated --enable-chunked-prefill flag and new connector for vllm v1! Am I mistaken that the "serve" here will cause a redundancy since the ENTRYPOINT is already ["vllm", "serve"]?

hickeyma replied:
Good catch, yes, its redundant. Updated `docker run` command to just pass the args to the `vllm serve` `entrypoint` in the image, i.e. `$HF_MODEL_NAME --kv-transfer-config`

sammshen commented on docker/Dockerfile:
```diff
@@ -1,16 +1,25 @@
-# docs/source/dev/dockerfile/dockerfile.rst and
-# docs/source/assets/dev/dockerfile-stages-dependency.png
+# Please update any changes made here to:
+# docs/source/developer_guide/docker_file.rst and
```
for the recently revamped the docs so I think these paths no longer exist.

would you mind editing to:
```text
# Instructions on installing lmcache/vllm-openai will be in:
# docs/source/getting_started/installation.rst
```

thank you!

hickeyma replied:
Updated

sammshen commented on docker/requirements.txt:
Could I please just clarify why the new `requirements.txt` file is smaller than the old `requirements-build.txt`? 
e.g. 
```text
filelock >= 3.10.4 # filelock starts to support `mode` argument from 3.10.4
partial-json-parser # used for parsing partial JSON outputs
```
are removed? 

Also, sorry if this is a dumb question, but you mentioned in https://github.com/LMCache/LMCache/issues/594#issuecomment-2887109824 consolidating the requirements. Is this generally best practice? If so, should we just merge them all in this PR as well? :)

hickeyma replied:
Good question. At this moment I am using [this requirements file](https://github.com/LMCache/LMCache/blob/dev/requirements.txt) as the file that defines the mandatory dependencies for LMCache.  I therefore replaced [requirements common](https://github.com/LMCache/LMCache/blob/dev/docker/requirements-common.txt) with this file. I have followed this approach to remove any unnecessary dependencies and from there identify if any dependencies are now missing.  I will add the diff below and maybe you can let me know if any dependencies need to be added back in and why?

hickeyma replied:
```command
$ diff requirements.txt requirements-common.txt

1,4d0
< # Copy of requirements.txt6 file in LMCache root
< # Remove this redundancy when add all requirement file to requirements directory in root
< 
< aiofiles
6c2,9
< infinistore
---
> einops # Required for Qwen2-VL.
> fastapi < 0.113.0; python_version < '3.9'
> fastapi >= 0.114.1; python_version >= '3.9'
> filelock >= 3.10.4 # filelock starts to support `mode` argument from 3.10.4
> gguf == 0.10.0
> importlib_metadata
> lm-format-enforcer == 0.10.6
> mistral_common >= 1.4.3
8,10c11,15
< nixl==0.2.0
< numpy
< nvtx
---
> numpy < 2.0.0
> openai >= 1.40.0 # Ensure modern openai package (ensure types module present)
> outlines >= 0.0.43, < 0.1
> partial-json-parser # used for parsing partial JSON outputs
> pillow  # Required for image processing
11a17,18
> prometheus-fastapi-instrumentator >= 7.0.0
> protobuf # Required by LlamaTokenizer.
12a20,21
> py-cpuinfo
> pydantic >= 2.9  # Required for fastapi >= 0.113.0
14,21c23,33
< pyzmq >= 25.0.0
< redis
< safetensors
< setuptools>=77.0.3,<80.0.0
< setuptools_scm>=8
< sortedcontainers
< torch==2.7.0 # Should correspond to the version used by latest vLLM package
< transformers >= 4.51.1
---
> pyzmq
> requests
> sentencepiece  # Required for LLaMA tokenizer.
> setuptools>=74.1.1; python_version > '3.11' # Setuptools is used by triton, we need to ensure a modern version is installed for 3.12+ so that it does not try to import distutils, which was removed in 3.12
> six>=1.16.0; python_version > '3.11' # transitive dependency of pandas that needs to be the latest version for python 3.12
> tiktoken >= 0.6.0  # Required for DBRX tokenizer
> tokenizers >= 0.19.1  # Required for Llama 3.
> tqdm
> transformers >= 4.45.0  # Required for Llama 3.2.
> typing_extensions >= 4.10
> uvicorn[standard]
```

hickeyma replied:
> Also, sorry if this is a dumb question, but you mentioned in https://github.com/LMCache/LMCache/issues/594#issuecomment-2887109824 consolidating the requirements. Is this generally best practice? If so, should we just merge them all in this PR as well? :)

Good question as well. I would prefer to do this in a follow on PR as it will mean changes across the project and not just within the context of the container image changes for this PR.

sammshen replied:
makes sense now! Since you already did the entire building and testing of the image, this sounds great and I personally don't know why the extra dependencies are there so we could just add them in the future if they are needed (e.g. one of them looks Qwen-specific?)

sammshen commented on docker/example_build.sh:
```diff
@@ -0,0 +1,19 @@
+
+# Update the following variables accordingly
+# Note: latest (built from GitHub) images of vLLM are in AWS ECR registry:
+# public.ecr.aws/q9t5s3a7/vllm-ci-test-repo
```
should lmcache also use ECR for latest images too or is DockerHub a perfectly viable option? I left a proposal about how to release images:
https://github.com/LMCache/LMCache/issues/594#issuecomment-2889322546

hickeyma replied:
@sammshen I think both are viable options. I'll do some investigation on this to understand this better.

hickeyma committed:
Update after review

Review comments:

- https://github.com/LMCache/LMCache/pull/665#discussion_r2094668070
- https://github.com/LMCache/LMCache/pull/665#discussion_r2094698460

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thank you @sammshen for the review and feedback. Updated and questions responded to.

Ready for review again.

sammshen commented:
LGTM!!

Siddhant-Ray reviewed (approved):
LGTM!

ApostaC commented on docker/example_run.sh:
```diff
@@ -1,14 +1,22 @@
+docker run --runtime=$RUNTIME --gpus all \
+    --env "HF_TOKEN=<HF_TOKEN_ACCESS_MODEL>" \
     --env "LMCACHE_USE_EXPERIMENTAL=True" \
     --env "chunk_size=256" \
     --env "local_cpu=True" \
     --env "max_local_cpu_size=5" \
```
Just realized this command is not correct. 
@hickeyma Can you do us a favor to quickly change the `--env` commands to the following: 
```bash
    --env "LMCACHE_CHUNK_SIZE=256" \
    --env "LMCACHE_LOCAL_CPU=True" \
    --env "LMCACHE_MAX_LOCAL_CPU_SIZE=5" \
```

Thanks!

(The full command can be found here: https://docs.lmcache.ai/developer_guide/docker_file.html)

ApostaC replied:
Actually NVM. I will create an issue for this

sammshen replied:
I can do this small fix rn!

hickeyma replied:
Good catch @ApostaC on the env variables. Thanks for raising issue #682

ApostaC reviewed (approved):
LGTM!
